File: sample-projects/php/src/Adapter/Persistence/PersistedCreatures.php
<?php

declare(strict_types=1);

namespace De\Sots\CellarsAndCentaurs\Adapter\Persistence;

require_once '../../Application/bootstrap.php';

use De\Sots\CellarsAndCentaurs\Application\CreatureFacade;
use De\Sots\CellarsAndCentaurs\Domain\Model\Creature;
use De\Sots\CellarsAndCentaurs\Domain\Model\CreatureId;
use De\Sots\CellarsAndCentaurs\Domain\Model\CreatureType;
use De\Sots\CellarsAndCentaurs\Domain\Model\NoSuchCreatureException;
use De\Sots\CellarsAndCentaurs\Domain\Service\Creatures;

final class PersistedCreatures implements Creatures
{
    public function __construct(private CreatureRepository $repository)
    {
    }

    public function save(Creature $creature): void
    {
        $this->repository->save(new CreatureEntity($creature->getId()->idAsString(), $creature->getType()->value));
    }

    public function find(CreatureId $id): Creature
    {
        $creatureEntity = $this->repository->findOne($id->idAsString());
        if ($creatureEntity === null) {
            throw new NoSuchCreatureException($id);
        }

        $type = CreatureType::tryFrom($creatureEntity->type) ?? CreatureFacade::STANDARD_CREATURE_TYPE;

        return new Creature(new CreatureId($creatureEntity->getId()), $type);
    }
}

File: sample-projects/rust/src/adapter/persistence/persisted_creatures.rs
use crate::adapter::persistence::creature_entity::CreatureEntity;
use crate::adapter::persistence::creature_repository::CreatureRepository;
use crate::application::CreatureFacade;
use crate::domain::model::creature::Creature;
use crate::domain::model::creature_id::CreatureId;
use crate::domain::model::no_such_creature_exception::NoSuchCreatureException;
use crate::domain::service::creatures::Creatures;

pub struct PersistedCreatures {
    repository: CreatureRepository,
}

impl PersistedCreatures {
    pub fn new(repository: CreatureRepository) -> Self {
        PersistedCreatures { repository }
    }
}

impl Creatures for PersistedCreatures {
    fn save(&mut self, creature: Creature) {
        self.repository.save(CreatureEntity::new(Some(creature.id().id.clone())));
    }

    fn find(&self, id: &CreatureId) -> Result<Creature, NoSuchCreatureException> {
        match self.repository.find_one(&id.id) {
            Some(entity) => Ok(Creature::with_type(
                CreatureId::new(entity.id.clone()),
                CreatureFacade::STANDARD_CREATURE_TYPE,
            )),
            None => Err(NoSuchCreatureException::new(id.clone())),
        }
    }
}
